nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,166 @@
+
+#include "nav2_map_server/map_mode.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: semantic_info_sub_(nullptr), map_filter_sub_(nullptr), map_filter_(nullptr)
+{
+}
+
+void KeepoutFilter::loadFilter(
+  const std::string semantic_info_topic)
+{
+  // Setting new semantic info for costmap filter subscriber
+  semantic_info_sub_ = node_->create_subscription<nav2_msgs::msg::CostmapFilterSemanticInfo>(
+    semantic_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::semanticInfoCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::semanticInfoCallback(
+  const nav2_msgs::msg::CostmapFilterSemanticInfo::SharedPtr msg)
+{
+  // Resetting previous subscriber each time when new semantic information arrives
+  if (map_filter_sub_) {
+    map_filter_sub_.reset();
+  }
+  map_filter_ = nullptr;
+
+  // Setting new map filter subscriber
+  map_filter_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    msg->map_filter_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::mapFilterCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::mapFilterCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  map_filter_ = msg;
+
+  // Updating costmap_filter_
+  unsigned int filter_cell_sz_x = map_filter_->info.width;
+  unsigned int filter_cell_sz_y = map_filter_->info.height;
+  double filter_map_res = map_filter_->info.resolution;
+  double filter_origin_x = map_filter_->info.origin.position.x;
+  double filter_origin_y = map_filter_->info.origin.position.y;
+  // ToDo - current Costmap2D does not support orientation
+  costmap_filter_ = std::make_unique<Costmap2D>(
+    filter_cell_sz_x, filter_cell_sz_y, filter_map_res, filter_origin_x, filter_origin_y);
+
+  // Fill the costmap_filter_ with values from map_filter_
+  unsigned int mx, my, it;
+  int8_t data;
+  for (mx = 0; mx < filter_cell_sz_x; mx++) {
+    for (my = 0; my < filter_cell_sz_y; my++) {
+      it = costmap_filter_->getIndex(mx, my);
+      data = map_filter_->data[it];
+      if (data == nav2_map_server::OCC_GRID_OCCUPIED) {
+        costmap_filter_->setCost(mx, my, LETHAL_OBSTACLE);
+      } else if (
+        data == nav2_map_server::OCC_GRID_FREE ||
+        data == nav2_map_server::OCC_GRID_UNKNOWN)
+      {
+        costmap_filter_->setCost(mx, my, NO_INFORMATION);
+      } else {
+        unsigned char cost = std::round(
+          data * (LETHAL_OBSTACLE - FREE_SPACE) / \
+          (nav2_map_server::OCC_GRID_OCCUPIED - nav2_map_server::OCC_GRID_FREE));
+        costmap_filter_->setCost(mx, my, cost);
+      }
+    }
+  }
+}
+
+void KeepoutFilter::process(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j,
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/)
+{
+  if (!map_filter_) {
+    RCLCPP_WARN(node_->get_logger(), "Map filter was not received");
```

Comment:
You should throttle this, it will get really annoying otherwise. 1hz or every 5 seconds or something would be fine

Reply:
Added message timer in https://github.com/ros-planning/navigation2/commit/fcf337566456677d0a6892b8539e902dfd086994#diff-f67a16397bb5979b7cbf4729ea53a451R100 (direct throttling like `rclcpp::sleep_for(>=2seconds)` will cause local planner to consider the plugin is dead).